lib/Validation.groovy

```diff
@@ -0,0 +1,132 @@
+import groovy.json.JsonSlurper
+import groovy.util.logging.Log
+
+class Validation {
+
+    /*
+    * Function to loop over all parameters defined in schema and check
+    * whether the given paremeters adhere to the specificiations
+    */
+    private static void validateParameters(params, json_schema, log, workflow){
+
+        def json = new File(json_schema).text
+        def Map json_params = (Map) new JsonSlurper().parseText(json).get('definitions')
+        def specified_param_keys = params.keySet()
+        def nf_params = ['profile', 'config', 'c', 'C', 'syslog', 'd', 'dockerize', 
+                        'bg', 'h', 'log', 'quiet', 'q', 'v', 'version']
+        def valid_params = []
+        def expected_params = []
+
+        // Loop over all parameters in schema and compare to given parameters
+        for (group in json_params){
+            for (p in group.value['properties']){
+                valid_params.push(validateParamPair(params[p.key], p, log))
+                expected_params.push(p.key)
+            }
+        }
+
+        // Check for nextflow core params and unexpected params
+        for (specified_param in specified_param_keys){
+            // nextflow params
+            if (nf_params.contains(specified_param)){
+                log.error "ERROR: That's a nextflow param!"
```

Comment:
Can force a system exit here I think. Happy to effectively make it an nf-core guideline that no parameter should share a name with a core Nextflow option.

Reply:
yep will add it there and if we get failing params as well I guess?

Reply:
Not for failing (unexpected?) params. There can be lots of reasons for params that are not in the schema (custom cluster options for example), so they should be warning only.

Could be good to collect them in a variable so that we can print them again if the workflow exits with an error?

I did also wonder a while back about checking for the number of characters different to valid params and failing if it's too close. Or could lowercase and remove all punctuation and check too for example. Could be nice to force an exit in those cases.

Reply:
Okay, yes I think it's good to give the user as much info as possible for debugging.
Yes a warning would be good in case it misses a valid parameter, will see about that.